### ISSUE TITLE
installing the latest npm with the npm module doesn't work

### ISSUE THREAD

**Issue by ansibot** [ansible/ansible#29192]
From @tkellen on 2014-12-07T15:31:41Z
##### Issue Type:

Bug Report
##### Component Name:

npm
##### Ansible Version:

ansible 1.7.1
##### Environment:

OSX 10.10
##### Summary:

Attempting to install the latest npm with the npm module doesn't seem to work.
##### Steps To Reproduce:

Put the following in a playbook where node is installed

``` yaml
- name: make sure the latest npm is installed
  npm: name=npm global=true
```

...or try running the following:
https://github.com/tkellen/ansible-vagrant-example/commit/903ee91e174fe4d444e248bc01d2d74b8ccf704c
##### Expected Results:

Latest version of npm is installed.
##### Actual Results:

NPM version is unchanged.

Copied from original issue: ansible/ansible-modules-extras#137

**Comment by ansibot** [comment]
From @tkellen on 2014-12-07T15:31:41Z

Can you really install npm with npm if npm isn't installed?  

Serious question... I need a bit more info here about how it was installed originally.

Can you paste the full verbose output here as well as your ansible version?

(Also, if you think you know how to fix this in the module, I'd be interested...not a node guy)

Thanks!

**Comment by ansibot** [comment]
From @tkellen on 2014-12-07T15:31:41Z

Thanks for the quick reply @mpdehaan!  The npm binary is installed by default alongside node, but is not always the most current version.  This line installs node and npm.
https://github.com/tkellen/ansible-vagrant-example/blob/master/ansible.yml#L14-L15

Once npm is installed, you can upgrade it by running `npm install -g npm`.

I don't have time to dig into this at the moment, my workaround is currently to do this:

``` yaml
- name: make sure the latest npm is installed
  command: npm install -g npm
```

**Comment by ansibot** [comment]
From @tkellen on 2014-12-07T15:31:41Z

If you pass the "global=yes" flag does it take effect?

**Comment by ansibot** [comment]
From @tkellen on 2014-12-07T15:31:41Z

:( nope

**Comment by ansibot** [comment]
From @tkellen on 2014-12-07T15:31:41Z

This is not an issue in at least ansible 1.9.1 and later. I think I spotted your mistake though; you are missing the state=latest parameter. The default value for state is present and of course if you are using npm it is guaranteed to be present so it will always be a no-op. If you add state=latest it will always update.

Tested by running: `ansible localhost -m npm -a "name=npm global=true state=latest"`

Without the state=latest, it did nothing as you reported, when I added the state param, it updated me to the latest version.

**Comment by ansibot** [comment]
From @tkellen on 2014-12-07T15:31:41Z

I have this kind of problem with ansible 1.9.2 too.

**Command**:

`ansible localhost -m npm -a "path=/path/to/my/dir/containing/package-json state=latest"`

**package.json**:

```
{
    "name": "Dummy",
    "version": "0.0.1",
    "devDependencies": {
        "grunt": "latest",
        "grunt-contrib-clean": "~0.6.0",
        "grunt-contrib-copy": "latest",
        "grunt-contrib-watch": "latest",
        "grunt-contrib-uglify": "latest",
        "grunt-contrib-compass": "latest",
        "grunt-contrib-cssmin": "~0.10.0",
        "grunt-contrib-concat": "~0.5.0"
    }
}
```

First run, everything is fine and I get `node_modules` directory, but then I update `package.json` file adding `"grunt-contrib-imagemin": "~0.4.0"` and call `ansible localhost -m npm -a "path=/path/to/my/dir/containing/package-json state=latest"` again, nothing updated. I always get

```
localhost | success >> {
    "changed": false
}
```

But if I run `npm install` shell command manually inside the directory, it works fine and I get my new plugin for imagemin.

**Comment by ansibot** [comment]
From @tkellen on 2014-12-07T15:31:41Z

@visay I think that's a separate issue worthy of its own ticket in that it is specifically about updates not working when using a package.json. The original title of this ticket does not relate to your question so it should have its own ticket to get noticed.

**Comment by ansibot** [comment]
From @tkellen on 2014-12-07T15:31:41Z

@feldoh OK, here we go #957

**Comment by ansibot** [comment]
From @tkellen on 2014-12-07T15:31:41Z

@tkellen I think this issue should be closed. As indicated by @feldoh, you're just missing 'state=latest' and the default is 'present', so it's working as expected.

**Comment by ansibot** [comment]
From @tkellen on 2014-12-07T15:31:41Z

See potentially-related issue (if the package isn't present already, `state=latest` doesn't install it currently): https://github.com/ansible/ansible-modules-extras/issues/1375

**Comment by ansibot** [comment]
From @tkellen on 2014-12-07T15:31:41Z

@geerlingguy Which btw should be addressed by PR #1663

**Comment by ansibot** [comment]
From @tkellen on 2014-12-07T15:31:41Z

@chrishoffman, ping. This issue is waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @tkellen on 2014-12-07T15:31:41Z

@shane-walker, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @tkellen on 2014-12-07T15:31:41Z

@shane-walker, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @tkellen on 2014-12-07T15:31:41Z

@shane-walker, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @tkellen on 2014-12-07T15:31:41Z

@shane-walker, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @tkellen on 2014-12-07T15:31:41Z

@shane-walker, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @xcambar
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by xcambar** [comment]
Hi,
Can the issue be reproduced with a recent ansible (2.4 or devel)?

As the new maintainer of the npm module, I am flooded by year-old issues. So as a triage strategy, any silent issue will be closed after 10 days unless you confirm your issue still exists in recent versions.

**Comment by visay** [comment]
@xcambar i can't reproduce it anymore with 2.4

**Comment by xcambar** [comment]
Great! Thanks for feedback!

close_me

**Comment by spruce** [comment]
It still persists for me (using ansible 2.4.3.0) when using it with https://github.com/geerlingguy/ansible-role-nodejs/
For instance I had an older version of `ghost-cli` installed. (1.4.1)

Via the mentioned roles method tried to update it to the latest version https://github.com/geerlingguy/ansible-role-nodejs/blob/a301eef42a4c6690fff3effd5130fa7ed0028fc3/tasks/main.yml#L26 with a variable set to 
```
nodejs_npm_global_packages:
  - name: ghost-cli
    version: latest
```
it reports:
```
ok: [blog.mysite] => (item={u'version': u'latest', u'name': u'ghost-cli'}) => {
    "changed": false,
    "invocation": {
        "module_args": {
            "executable": null,
            "global": true,
            "ignore_scripts": false,
            "name": "ghost-cli",
            "path": null,
            "production": false,
            "registry": null,
            "state": "latest",
            "version": "latest"
        }
    },
    "item": {
        "name": "ghost-cli",
        "version": "latest"
    }
}
```
but when running the task:
```
- name: make sure the latest npm is installed
  command: npm install -g ghost-cli@latest
```

it reports with many lines (just some npm stuff saying it updated to 1.5.2 the currently latest version)

**Comment by kimegede** [comment]
I can reproduce this too.

```
---

- name: Install foo/bar from private registry
  become: yes
  npm:
    name: "@foo/bar"
    registry: 'http://npm.private-registry.example/'
    global: yes
    state: latest
```

* 1st. time it's installed (eg. v.2.0.0 is installed)
* 2nd. time nothing changed (which is good)
* If I manually installing an older version `npm install -g @foo/bar@v1.0.0`
* Running Ansible again, it says changed - but it haven't installed the latest version.
I would have expected it to update it, I have tried to add `version: latest` as well - same outcome.

**Comment by kimegede** [comment]
EDIT: The above seems to be because of an old version of npm (`4.6.1`) where `npm update` doesn't update (at least not what is needed).
The feature works from `5.2.0`+

**Comment by pgrenaud** [comment]
@spruce I having the same issue, but I found a work around. Removing the line `version: latest` and using only `state: latest` fix the issue. Here's an example:

Before, my task looked like this. Ansible didn't report any changes and no package were being updated.
```yaml
- name: update npm global packages
  npm:
    name: '{{ item }}'
    version: latest
    global: yes
    state: latest
  environment:
    NPM_CONFIG_USER: nobody
    NPM_CONFIG_GROUP: nogroup
  with_items: '{{ npm_packages }}'
  tags:
    - always
```

After, my task looked like this. Ansible did report the changes and packages were updated.
```yaml
---
- name: update npm global packages
  npm:
    name: '{{ item }}'
    global: yes
    state: latest
  environment:
    NPM_CONFIG_USER: nobody
    NPM_CONFIG_GROUP: nogroup
  with_items: '{{ npm_packages }}'
  tags:
    - always
```

@xcambar So, there seems to be an issue with the parameter `version`, interfering with the `state` parameter. The docs are more or less clear about this, but I believe that if both `version` and `state` parameters are set to `latest`, the packages should be updated.

Also, while testing, I observe a single package, `gulp-cli`, not being updated even if Ansible did report the package as changed (therefore updated). I ran the task multiple time with the same result. I had to manually update this package (while all my other packages were successfully updated with Ansible).
